### PR TITLE
Fix QuantityType parsing for exponential numbers with units

### LINKED ISSUE #5247: [Weather Underground] Error invoking #valueOf(String)
OH 2.3.0 snapshot 1231
WU binding 0.10.0.201803161254

I don't know if this is an issue with the binding, [Units of Measurement](https://github.com/eclipse/smarthome/issues/5243), or my setup. All of my other weather items using UOM are working, but two of my items...
```
Number:Speed   Weather_Wind_Speed_Mph       "Wind Speed [%.0f mph]"	        <wind>	(gWeather)  {channel="weatherunderground:weather:71ed51c0:current#windSpeed"}
Number:Speed   Weather_Wind_Gust_Mph        "Wind Speed (gust) [%.0f mph]"	<wind>	(gWeather)  {channel="weatherunderground:weather:71ed51c0:current#windGust"}
```

... are getting values and displaying properly...
image

...but appear to be occasionally causing an error:
```
2018-03-18 00:07:36.535 [ERROR] [clipse.smarthome.core.internal.events.EventHandler] - Creation of ESH-Event failed, because one of the registered event factories has thrown an exception: Error invoking #valueOf(String) on class 'org.eclipse.smarthome.core.library.types.QuantityType' with value '0E-22 mph'.
java.lang.IllegalStateException: Error invoking #valueOf(String) on class 'org.eclipse.smarthome.core.library.types.QuantityType' with value '0E-22 mph'.
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.parseSimpleClassName(ItemEventFactory.java:178) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.parseType(ItemEventFactory.java:146) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.getState(ItemEventFactory.java:124) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.createStateEvent(ItemEventFactory.java:111) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.createEventByType(ItemEventFactory.java:77) ~[?:?]
        at org.eclipse.smarthome.core.events.AbstractEventFactory.createEvent(AbstractEventFactory.java:50) ~[?:?]
        at org.eclipse.smarthome.core.internal.events.EventHandler.createESHEvent(EventHandler.java:121) ~[?:?]
        at org.eclipse.smarthome.core.internal.events.EventHandler.handleEvent(EventHandler.java:95) ~[?:?]
        at org.eclipse.smarthome.core.internal.events.EventHandler.handleEvent(EventHandler.java:72) ~[?:?]
        at org.eclipse.smarthome.core.internal.events.ThreadedEventHandler.lambda$0(ThreadedEventHandler.java:71) ~[?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.GeneratedMethodAccessor321.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.parseSimpleClassName(ItemEventFactory.java:169) ~[?:?]
        ... 10 more
Caused by: java.lang.IllegalArgumentException: E not recognized (in 0 E-22 mph at index 2)
        at tec.uom.se.quantity.Quantities.getQuantity(Quantities.java:80) ~[?:?]
        at org.eclipse.smarthome.core.library.types.QuantityType.<init>(QuantityType.java:78) ~[?:?]
        at org.eclipse.smarthome.core.library.types.QuantityType.valueOf(QuantityType.java:121) ~[?:?]
        at sun.reflect.GeneratedMethodAccessor321.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.parseSimpleClassName(ItemEventFactory.java:169) ~[?:?]
        ... 10 more
```

### PR DESCRIPTION
**Pull request by htreu**
Fixes #5247.

Signed-off-by: Henning Treu <henning.treu@telekom.de>

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/QuantityType.java [hunk 1]
```diff
@@ -74,7 +75,8 @@ public QuantityType(String value) {
         String[] constituents = value.split(UNIT_PATTERN);
 
         // getQuantity needs a space between numeric value and unit
-        String formatted = String.join(" ", constituents);
+        constituents = (String[]) ArrayUtils.add(constituents, constituents.length - 1, " ");
```
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/QuantityType.java]: If I understood it correctly, the problem was that the `UNIT_PATTERN` pattern matched twice, so what with the exponent we ended up having 3 constituents, right? How can we be sure then that `length - 1` is _always_ the right place to separate the numeric part from the units part?

Comment (htreu) [reply bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/QuantityType.java]: We assume that the unit is always the last part of the constructor String. The old implementation just did that by joining the two constituent by `" "`.
I'm not aware of other notation, but right now we would not support them.

**Comment by maggu2810** [comment]
@sjka You are too fast :wink: 
I am still thinking about that lines:

```Java
        String[] constituents = value.split(UNIT_PATTERN);

        // getQuantity needs a space between numeric value and unit
        constituents = (String[]) ArrayUtils.add(constituents, constituents.length - 1, " ");
        String formatted = String.join("", constituents);
```

How much elements could `constituents` consist of?

There is Apache Commons used and an unchecked cast (at least a suppressed warning is used). Can we solve that without a third party lib and with a strict type safe system?

**Comment by htreu** [comment]
@maggu2810 yes we can. I had a version previously on record which just lopped through the String array filling a StringBuilder and adding the `" "` before the last element.
It was 4 more lines but could be refactored into a private helper method. wdyt?

**Comment by maggu2810** [comment]
And another question, are you sure, the implementation is correct?

If I use something like that:
```Java
        QuantityType<Dimensionless> dt0 = new QuantityType<>("12E-2");
```

I got this one:
```text
java.lang.IllegalArgumentException: E not recognized (in 12 E-2 at index 3)
	at tec.uom.se.quantity.Quantities.getQuantity(Quantities.java:80)
	at org.eclipse.smarthome.core.library.types.QuantityType.<init>(QuantityType.java:86)
	at org.eclipse.smarthome.core.library.types.QuantityTypeTest.testExponentials(QuantityTypeTest.java:239)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)
```

**Comment by htreu** [comment]
@maggu2810 thanks for the Stacktrace. It actually already fails when parsing the quantity. I will think about it and provide a fix.

**Comment by sjka** [comment]
> @sjka You are too fast 😉

Sorry!

**Comment by maggu2810** [comment]
@htreu 

```Java
    /**
     * Creates a new {@link QuantityType} with the given value. The value may contain a unit. The specific
     * {@link Quantity} is obtained by {@link Quantities#getQuantity(CharSequence)}.
     *
     * @param value the non null value representing a quantity with an optional unit.
     */
    @SuppressWarnings("unchecked")
    public QuantityType(String value) {
        final Matcher matcher = Pattern.compile("(-?[0-9\\.]*)(E[+-]?[0-9]+)?\\s*(.*)").matcher(value);
        final String formatted;
        if (matcher.matches()) {
            // matcher.group(1) => value
            // matcher.group(2) => exponent, optional
            // matcher.group(3) => unit
            if (matcher.group(2) == null) {
                formatted = String.format("%s %s", matcher.group(1), matcher.group(3));
            } else {
                formatted = String.format("%s%s %s", matcher.group(1), matcher.group(2), matcher.group(3));
            }
        } else {
            formatted = value;
        }

        quantity = (Quantity<T>) Quantities.getQuantity(formatted);
    }
```

**Comment by htreu** [comment]
Sorry @maggu2810 I saw your solution too late.
I managed to refine the regex so it now matches (splits) on the correct position. See #5267.

**Pull request merged**